File: app/src/main/java/de/bikebean/app/ui/drawer/log/GithubGistUploader.java
package de.bikebean.app.ui.drawer.log;

import android.content.Context;
import android.os.AsyncTask;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.IOException;

import de.bikebean.app.R;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class GithubGistUploader extends AsyncTask<String, Void, Boolean> {

    private final LogViewModel mLogViewModel;
    private final @NonNull UploadSuccessNotifier mUploadSuccessNotifier;

    private final @NonNull String mDescription;
    private final @NonNull String mSmsTsv;
    private final @NonNull String mStateTsv;
    private final @NonNull String mLogTsv;

    private final @NonNull String mUrl;
    private final @NonNull String githubGistsToken;

    public interface UploadSuccessNotifier {
        void notifyUploadSuccess(boolean success);
    }

    public GithubGistUploader(final @NonNull Context context, LogViewModel lv,
                              final @NonNull UploadSuccessNotifier usn,
                              final @NonNull String description, final @NonNull String smsTsv,
                              final @NonNull String stateTsv, final @NonNull String logTsv) {
        mLogViewModel = lv;
        mUploadSuccessNotifier = usn;

        mDescription = description;
        mSmsTsv = smsTsv;
        mStateTsv = stateTsv;
        mLogTsv = logTsv;

        githubGistsToken = context.getString(R.string.github_gist_token);
        mUrl = context.getString(R.string.github_gists_baseurl);
    }

    @Override
    protected @NonNull Boolean doInBackground(final @NonNull String... args) {
        final @Nullable MediaType mediaType = MediaType.parse("application/json");

        final @NonNull GithubGistBody jsonBody =
                new GithubGistBody(mDescription, mSmsTsv, mStateTsv, mLogTsv);
        final @NonNull RequestBody body =
                RequestBody.create(mediaType, jsonBody.createJsonApiBody());

        final @NonNull OkHttpClient client = new OkHttpClient().newBuilder()
                .build();
        final @NonNull Request request = new Request.Builder()
                .url(mUrl)
                .method("POST", body)
                .addHeader("Authorization", "token " + githubGistsToken)
                .addHeader("Content-Type", "application/json")
                .build();

        final @NonNull Response response;

        try {
            response = client.newCall(request).execute();
            mLogViewModel.d("Successfully uploaded Crash Log (" + response.code() + ")");
            return true;
        } catch (IOException e) {
            mLogViewModel.e("Could not upload Crash Log: " + e.toString());
            return false;
        }
    }

    @Override
    protected void onPostExecute(final @NonNull Boolean isLogUploaded) {
        mUploadSuccessNotifier.notifyUploadSuccess(isLogUploaded);
    }
}

File: app/src/main/java/de/bikebean/app/ui/drawer/log/GithubGistUploader.kt
package de.bikebean.app.ui.drawer.log

import android.content.Context
import de.bikebean.app.R
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException

class GithubGistUploader(
        context: Context,
        private val mLogViewModel: LogViewModel,
        private val mUploadSuccessNotifier: (Boolean) -> Unit,
        private val mDescription: String,
        private val mSmsTsv: String,
        private val mStateTsv: String,
        private val mLogTsv: String) {

    private val mUrl: String = context.getString(R.string.url_github_gists)
    private val githubGistsToken: String = context.getString(R.string.github_gist_token)

    fun execute() {
        CoroutineScope(Dispatchers.Main).launch {
            mUploadSuccessNotifier(doInBackground())
        }
    }

    private suspend fun doInBackground(): Boolean =
        withContext(Dispatchers.IO) {
            try {
                val response: Response = OkHttpClient()
                        .newBuilder()
                        .build()
                        .newCall(buildRequest())
                        .execute()

                if (response.isSuccessful) {
                    mLogViewModel.d("Successfully uploaded Crash Log (${response.code})")
                    true
                } else {
                    mLogViewModel.d("Could not upload Crash Log (${response.code}): "
                            + "${response.body?.string()}"
                    )
                    false
                }
            }

            catch (e: IOException) {
                mLogViewModel.e("Could not upload Crash Log: $e")
                false
            }
        }

    private fun buildRequest(): Request {
        val request = GithubGistBodyBuilder(mDescription, mSmsTsv, mStateTsv, mLogTsv)
                .build()
                .toRequestBody(mediaType)

        return Request.Builder()
                .url(mUrl)
                .method("POST", request)
                .addHeader("Authorization", "token $githubGistsToken")
                .addHeader("Content-Type", "application/json")
                .build()
    }

    companion object {
        val mediaType = "application/json".toMediaTypeOrNull()
    }
}